### ISSUE TITLE
`DAG.to_pdag()` not implemented

### ISSUE THREAD

**Issue by behrman** [pgmpy/pgmpy#1625]
Currently `DAG.to_pdag()` is stub that always returns `None`. This could confuse users. It would probably be best to leave the method out completely until it has been implemented.

The dagitty package provides this functionality with `equivalenceClass()` ([doc](https://cran.r-project.org/web/packages/dagitty/dagitty.pdf), pp. 13 - 14). The code that implements this is the JavaScript function `GraphTransformer.dagToCpdag()` ([code](https://github.com/jtextor/dagitty/blob/master/jslib/graph/GraphTransformer.js), lines 1167 - 1184) with unit tests ([code](https://github.com/jtextor/dagitty/blob/0686b40e007919c949caa6e98706389f0f68a62a/test/test/graph-transformations.js), lines 64 - 69).

@ankurankan

**Comment by ankurankan** [comment]
@behrman Thanks for reporting this. I will remove the method for now till it is implemented.

**Comment by behrman** [comment]
@ankurankan If you can remember, please comment to this thread when `DAG.to_pdag()`has been implemented. I will update the [pgmpy tutorial](https://github.com/behrman/cis/blob/master/solutions_pgmpy.ipynb) I've written to include an illustration of its use.

**Comment by parthrastogicoder** [comment]
Can I contribute in making of the function  DAG.to_pdag() ?

**Comment by ankurankan** [comment]
@parthrastogicoder Yeah, that would be great. Feel free to reach out to me if you need any help with it.

**Comment by Nuna7** [comment]
@ankurankan Hi ankur, I look at this issue. 

It seems like in the previous constructor of PDAG, cycles are detected and you use the following code to filter out the undirected edges :

```python
# Filter out undirected edges as they also form a cycle in
# themself when represented using directed edges.
cycles = filter(lambda t: len(t) > 2, nx.simple_cycles(self))
if cycles:
     out_str = "Cycles are not allowed in a PDAG. "
```

But when searching for cycles on directed edges (same as filtering out undirected edges ) using the following code, no cycle is detected.

```python
super(PDAG, self).__init__(
            directed_ebunch
            + undirected_ebunch
            + [(Y, X) for (X, Y) in undirected_ebunch]
        )
self.latents = set(latents)
self.directed_edges = set(directed_ebunch)
self.undirected_edges = set(undirected_ebunch)

directed_graph = nx.DiGraph(self.directed_edges)
cycles = list(nx.simple_cycles(directed_graph))
            
if cycles:
      out_str = "Cycles are not allowed in a PDAG. "
```

Further, when checking for cycles by filtering out the undirected edges as you did, it seems like the following condition
```python
cycles = filter(lambda t: len(t) > 2, nx.simple_cycles(self))
if cycles:
      out_str = "Cycles are not allowed in a PDAG. "
```
is only true because the return object of the filter method is **<filter object at 0x7fd372abeeb0>** while if we transform this object into **list**,  there is no cycle detected.

```python
cycles = filter(lambda t: len(t) > 2, nx.simple_cycles(self))
print(cycles)
print(list(cycles))
 ```

```bash
<filter object at 0x7fd372abeeb0>
[]
```

and thereby even when we just change the condition to 

```python
if list(cycles):
      out_str = "Cycles are not allowed in a PDAG. "
```
, the constructor didn't detected any cycles (no more outputting the string which said that cycles are not allowed).

**Comment by ankurankan** [comment]
@Nuna7 I am not sure if the cycle issue is relevant to this issue. Maybe you are referring to #1300 ? (or am I misunderstanding something?)

**Comment by Nuna7** [comment]
@ankurankan  Oh, sorry for commenting on wrong place. I will remove my comment. Further, it seem like the problem is still not solve in #1300 due to some failed test case, I am only try it in some other approach, which pass all test such as the commented test case in test_DAG.py, should I make pull or comment in #1300 and tell my approach ?

**Comment by Manas-7854** [comment]
Hi @ankurankan, i was thinking of taking up this issue, can you tell me `GraphAnalyzer.isEdgeStronglyProtected` inside pgmpy?

**Comment by ankurankan** [comment]
@Manas-7854 No, that is not implemented in pgmpy. You can however, find the full algorithm for converting DAG to CPDAG here: https://www.jmlr.org/papers/volume2/chickering02a/chickering02a.pdf#page=24.70 (Page 455; Figure 4 and 5)

**Comment by dhruvd-1** [comment]
hi can this be assigned to me pls

**Comment by Vidhaankhare16** [comment]
Hello @ankurankan I have read the concerned docs and I think I will be able to work on this as my first issue , if no one is working on this currently can I please be assigned?

**Comment by ankurankan** [comment]
Already implemented in #2068
